### ISSUE TITLE
ionic2 - enhancement(transition) - Give the nav controller's transition a finish event

### ISSUE THREAD

**Issue by zakton5** [driftyco/ionic#8238]
#### Short description of the problem:
There are many cases where I need to wait on a push's transition to finish before doing something on the pushed page. For example:

I start on Page1, push Page2, and then need to display some kind of toast or alert immediately once the Page2 AND its data have finished loading. The problem here is that when my data has loaded, I do not know if the page's transition has finished yet, and vice versa.

#### What behavior are you expecting?
It would be great if I could do something like this on Page2:
```ts
this.navCtrl.onTransitionFinished(() => {
    this.toastCtrl.create({
        ...
    });
});
```
This would be very convenient.

**Comment by NoNameProvided** [comment]
The `ionViewDidEnter` function is what you are looking for. From the [docs](http://ionicframework.com/docs/v2/api/components/nav/NavController/):

> Runs when the page has fully entered and is now the active page. This event will fire, whether it was the first load or a cached page.

**Comment by comfortme** [comment]
like this?

> let navTransitionPromise = this.navController.push(Page2);
navTransitionPromise.then( () => {
  // the transition has completed, so I can push another page now
  return this.navController.push(Page3);
}).then( () => {
  // the second transition has completed, so I can push yet another page
    return this.navController.push(Page4);
}).then( () => {
  console.log('The transitions are complete!');
})

**Comment by zakton5** [comment]
@NoNameProvided The ionViewDidEnter function is often sufficient because it runs after the transition is finished. But at other times it is not sufficient because the ionViewDidEnter function runs every single time the page is entered.

So using ionViewDidEnter: if I have Page1, and push Page2 (where the alert should appear) it will appear correctly the first time. Then, I push Page3, and then pop Page3. The alert will appear again, even though I only wanted it to appear the first time.

@comfortme That chain of promises is all in the same controller. There is no way for Page1 to truly know when all of Page2's data has finished loading. What I am looking for is this on Page1:

```ts
this.navCtrl.push(Page2);
```

and this on Page2
```ts
constructor() { // or some other init function
    this.dbService.loadData().then(() => { // All data needs to load first
        this.navCtrl.transitionFinished(() => {
            // Dismiss loading, Show alert/toast etc.
        }):
    });
}
```


Basically, it would be nice for the _pushed_ page to know when its *first* transition has finished. If there is some nice workaround for this that is fine too.

**Comment by NoNameProvided** [comment]
With this you simply push 3 pages to the stack after each other, but in your first post you asked about a method to detect when the entering page is fully visible and ready for use. 

 `ionViewDidEnter` needs to be added to the entering page like 

```ts
@Component({
  templateUrl: 'build/pages/page2/page2.component.html',
  styleUrls: ['build/css/pages/page2/page2.component.css']
})
export class Page2 {
  constructor() {
    console.log('The constructor start to run before the navigation is started.'); 
  }

  ionViewDidEnter() {
    console.log('Entering transition of Page2 has been finished. You securely pop your toast messages now.');
    this.toastCtrl.create({
        // ...
    });
  }
}
```
And then you simply push the page with `NavController.push(Page2)` from `Page1`.
 

What you have wrote in your second comment will work. It will push the pages after each other to the stack always waiting to the previous one's animations to finish. But I don't see why you would need to do that, if you want to push more than one page to the stack you can use the `NavController.setPages([Page2, Page3, Page4])` [method](https://ionicframework.com/docs/v2/api/components/nav/NavController/#setPages). This will trigger only one animation. (`Page4` will be entered with that animation, but in the background the other pages will be added to the stack as well.)

**Comment by zakton5** [comment]
I understand that, but my specific use case doesn't work with this solution.
I have Page1, and push Page2, which is a form that allows the user to save an issue or problem they have to the DB. However, if there are already issues in the DB, then an alert needs to appear telling the user so that they do not create a duplicate issue. So the alert appears (which is done in ionViewDidEnter) and the user then navigates to Page3 to view these existing issues. Once the user is finished, he/she navigates back to Page2. At this point, I do not want the alert to appear again because that would be redundant. But because that functionality is in ionViewDidEnter, it will appear anyway. Is there a way around this?

**Comment by jgw96** [comment]
Hello, thanks for using Ionic! @zakton5 it sounds like `ionViewDidLoad` is what you are looking for. This lifecycle event only runs the first time the component is loaded. Also, since this seems more like a support question instead of a bug with Ionic I am going to be closing this issue and pointing you towards our [forum](https://ionicframework.forum.com). Thanks!

**Comment by zakton5** [comment]
Unfortunately ```ionViewDidLoad``` is not quite what I want. That method runs _before_ the transition is finished. I want a way to capture when the transition is finished that is not ```ionViewDidEnter```

**Comment by NoNameProvided** [comment]
@zakton5 When you push to `Page2` pass a parameter object like `this.nav.push(Page2, { showAlert: true })` This is accessible through the `NavParams` class only when the page is pushed. So when you pop back from `Page3` your alert won't be showed.

Minimal example: 

```ts
@Component({
  templateUrl: 'build/pages/page2/page2.component.html',
  styleUrls: ['build/css/pages/page2/page2.component.css']
})
export class Page2 {
  constructor(private params: NavParams) {
    console.log('The constructor start to run before the navigation is started.'); 
  }

  ionViewDidEnter() {
    console.log('Entering transition of Page2 has been finished. You securely pop your toast messages now.');
    if(this.params.get('showAlert')) {
      this.toastCtrl.create(/* ... */);
    }
  }

}
```

**Comment by zakton5** [comment]
Yeah that solution will work. Thanks!
